Introduce Application Privileges to Roles

tvernum opened the pull request:
This commit introduces "Application Privileges" (aka custom privileges) to the X-Pack security model.

Application Privileges are managed within Elasticsearch, and can be tested with the `_has_privileges` API, but do not grant access to any actions or resources within Elasticsearch.
Their purpose is to allow applications outside of Elasticsearch to represent and store their own privileges model within Elasticsearch roles.

Specifically, this adds
 - GET/PUT/DELETE actions for defining application level privileges
 - application privileges in role definitions
 - application privileges in the has_privileges API

Closes: #29820

elasticmachine commented:
Pinging @elastic/es-security

kobelb commented:
This is awesome, thanks so much @tvernum. I just started integration it with our stuff, and I'll let you know how it progresses.

I noticed that users with the `superuser` role don't intrinsically get all application privileges, would it be possible to change this so they do?

tvernum commented:
> I noticed that users with the superuser role don't intrinsically get all application privileges, would it be possible to change this so they do?

Thanks for the reminder, I had forgotten about that.

kobelb commented:
Also, in the `_has_privileges` call, it would seem that we'd want to be using "actions" the key in the application array, as opposed to "privileges", as we're checking to see if they have the specified actions.

kobelb commented:
You might be planning to explain this in the docs, but would you mind explaining the role that the `:` plays when defining the actions for the custom privileges? I was able to create the following privilege/actions:

```
{
  "kibana": {
    "read": {
      "application": "kibana",
      "name": "read",
      "actions": [
        "version:7.0.0-alpha1",
        "saved-objects/dashboard/get",
      ],
      "metadata": {}
    }
  }
}
```

but then when I went to check the _has_privileges with

```
{
    "applications":[
        {
            "application":"kibana",
            "resources":["*"],
            "privileges":["saved-objects/dashboard/get"]
        }
    ]
}
```

it was returning false.

I noticed that I probably should be using the `action:` prefix so that I can always be doing the version check properly, but I feel like I'm missing something behind how all these components work together.

kobelb commented:
Apologies for all the questions, does Elasticsearch do some type of caching with the application privileges? I've noticed a few situations where things were behaving weirdly and not authorizing users as I expected where if I restarted Elasticsearch and ran the steps again, it's begin working again. It might just be a "user error" as well.

tvernum commented:
>  it would seem that we'd want to be using "actions" the key in the application array, as opposed to "privileges", as we're checking to see if they have the specified actions.

We _could_, but I've tried to mimic the `cluster` and `index` fields in `has_privileges` which use `privileges`.
The "privileges" field accepts either actions or high-level-privilege names. For your purposes it only makes sense to include action names, but the field allows a mix of both.

> would you mind explaining the role that the : plays when defining the actions for the custom privileges

The is supposed to be anything to explain, and I don't understand why you saw the behaviour you did. 
The only rule is that we need some way to distinguish between privilege names and action names, and we do that with `/` or `:` (or `*`, but that's actually playing the part of a wildcard).
You definitely shouldn't see any difference between `/` and `:`.

> does Elasticsearch do some type of caching with the application privileges?

There shouldn't be any. There's caching on roles, but I pulled out what I think we the last piece of caching on privielges because it was broken.

tvernum commented:
> I don't understand why you saw the behaviour you did.

I haven't been able to reproduce this. I suspect it's related to whatever caching issue you saw (whether real or user-error), but if you can still reproduce it let me know.

My test case is:
**Privileges**
```
POST /_xpack/security/privilege
{"kibana": {
   "login":{"application":"kibana","name":"login","actions":["action:login","version:6.4.*"]},
   "read":{"application":"kibana","name":"read","actions":["action:login","version:6.4.*","data:read/*","saved-objects/dashboard/get"]}
}}
```
**Role**
```
POST /_xpack/security/role/test
{"applications":[{"application":"kibana","privileges":["read"],"resources":["*"]}]}
```
**User**
```
POST /_xpack/security/user/test
{ "password" : "magic123", "roles" : [ "test" ] }
```
**Check**
```
GET /_xpack/security/user/_has_privileges
{
  "index": [],  
  "applications": [ { "application": "kibana", "resources":["test"], "privileges":["saved-objects/dashboard/get"] } ]
}
# ...
{
  "username" : "test",
  "has_all_requested" : true,
  "cluster" : { },
  "index" : { },
  "application" : {
    "kibana" : {
      "test" : {
        "saved-objects/dashboard/get" : true
      }
    }
  }
}
```

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/ApplicationPermission.java:
```diff
@@ -0,0 +1,83 @@
+                return new PermissionEntry(k, Automatons.unionAndMinimize(Arrays.asList(existing.resources, patterns)));
+            }
+        }));
+        this.permissions = new ArrayList<>(permissionsByPrivilege.values());
```
make this an unmodifiable list?

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/ApplicationPermission.java:
```diff
@@ -0,0 +1,83 @@
+    public boolean grants(ApplicationPrivilege other, String resource) {
+        Automaton resourceAutomaton = Automatons.patterns(resource);
+        final boolean matched = permissions.stream().anyMatch(e -> e.grants(other, resourceAutomaton));
+        logger.debug("Permission [{}] {} grant [{} , {}]", this, matched ? "does" : "does not", other, resource);
```
this feels like a trace level log to me

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/ApplicationPermission.java:
```diff
@@ -0,0 +1,83 @@
+        this.permissions = new ArrayList<>(permissionsByPrivilege.values());
+    }
+
+    public boolean grants(ApplicationPrivilege other, String resource) {
```
Can you add javadocs?

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java:
```diff
@@ -0,0 +1,268 @@
+    }
+
+    public ApplicationPrivilege(String application, String privilegeName, String... patterns) {
+        this(application, Sets.newHashSet(privilegeName), patterns, Collections.emptyMap(), true);
```
use Collections.singleton here

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java:
```diff
@@ -0,0 +1,268 @@
+        this(application, Sets.newHashSet(privilegeName), patterns, Collections.emptyMap(), true);
+    }
+
+    private ApplicationPrivilege(String application, Set<String> name, Collection<String> patterns, Map<String, Object> metadata,
```
do we need this constructor?

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/DeletePrivilegesRequestBuilder.java:
```diff
@@ -0,0 +1,38 @@
+        this(client, DeletePrivilegesAction.INSTANCE);
+    }
+
+    public DeletePrivilegesRequestBuilder(ElasticsearchClient client, DeletePrivilegesAction action) {
```
Unfortunately need it in order to implement `DeletePrivilegesAction.newRequestBuilder`, but I've made it package protected.

kobelb commented:
@tvernum you were right that it appears to be a caching issue. If I perform the following steps to create the privileges, and a role/user with those privileges, and then update the privileges, the _has_privileges API returns false until I clear the cache for that role:

```
POST _xpack/security/privilege
{
  "kibana": {
    "read":{
      "application":"kibana",
      "name":"read",
      "actions":[
        "version:7.0.0-alpha1",
        "action:saved-objects/config/get"
      ]
    }
  }
}
```

```
POST _xpack/security/role/kibana_rbac_user
{
  "cluster": [],
  "indices": [],
  "applications": [
    {
      "application": "kibana",
      "privileges": [ "read" ],
      "resources": [ "default" ]
    }
  ]
}
```

```
PUT _xpack/security/user/brandon
{
  "password" : "password",
  "roles" : [ "kibana_rbac_user" ],
  "full_name" : "Brandon",
  "email" : "someone@gmail.com"
}
```

```
POST _xpack/security/user/_has_privileges
{
    "applications":[
        {
            "application":"kibana",
            "resources":["default"],
            "privileges":["action:saved-objects/config/get"]
        }
    ]
}
```

^^ this returns `has_all_requested: true`

```
POST _xpack/security/privilege
{
  "kibana": {
    "read":{
      "application":"kibana",
      "name":"read",
      "actions":[
        "version:7.0.0-alpha1",
        "action:saved-objects/config/get",
        "action:saved-objects/config/search"
      ]
    }
  }
}
```

```
POST _xpack/security/user/_has_privileges
{
    "applications":[
        {
            "application":"kibana",
            "resources":["default"],
            "privileges":["action:saved-objects/config/search"]
        }
    ]
}
```

^^ this returns `has_all_requested: false`

Is this expected behavior? We're planning on occasionally updating the privileges map, and having to clear the cache of every role that has those privileges is rather cumbersome.

tvernum commented:
> Is this expected behavior? 

It wasn't something I predicted, but it's consistent with some other caches in security.
If you change role-mappings it doesn't affect users that already cached. 

Updates to roles are live because the cached link between Users and Roles is simply by name, so if the role changes in the cache the User will see the new role on next request.

But if you change a role-mapping so that LDAP user "bob" is supposed to now have "kibana_user", that won't be effective until the cache for "bob" is cleared.

> We're planning on occasionally updating the privileges map, and having to clear the cache of every role that has those privileges is rather cumbersome.

You _can_ clear the cache for all roles in 1 API call, though obviously that's a pretty blunt instrument.

I'll look at options, but my gut feel is that I'd like to update the clear roles API to be able to accept an application name so you can clear all roles that have "kibana" privileges once you've finished adding/deleting/updating privileges.

tvernum commented:
> We're planning on occasionally updating the privileges map,

@kobelb How often does this happen? We can change the clear-cache API to ben more granular, but if the updates are infrequent, I think it would be simpler just to clear the whole cache when you update.

kobelb commented:
@tvernum the current plan is to execute the Privileges PUT every time that an instance of Kibana starts up for all the custom privileges to ensure that we have the information in Elasticsearch to perform the authorization using the current privileges.

In an ideal configuration, this would only happen once for every instance of Kibana for a long period of time; however, in an unideal situation or one where new instances were being provisioned frequently, it could happen somewhat frequently. Unfortunately, *it depends*...

tvernum commented:
> will a more granular control of the cache have any big impact on the performance of the API?

My concern is more about the design and complexity of the clear cache API, rather than performance. 
The change requires that we can somehow express a variety criteria about which roles need to be cleared - at the moment it's _all_ or by-name. This would add another option.

My choices to do that are:
- something relatively small and bespoke, which would produce a weird API where the set of things you can express aligns to what Kibana needs right now, but doesn't plan for anything else, and locks the API into that behaviour at least until 7.0
- something larger and more cleanly designed, which adds a heap more complexity just to solve a relatively small problem.

I don't _like_ either of those options - they both place a high price on future maintenance that I would love to avoid if that's at all possible. It might not be possible, but we need to be sure of that.

kobelb commented:
> Can you not check whether this is neccessary before doing it? The GET API will allow you to check what's present, and you can use the per-privilege metadata to store a version number, or build hash, or something to compare against. Then it can be just a case of comparing that the privilege names match and the metadata.version matches.
>That's what we do inside ES security - before we update the index mappings, we check the metadata version.
> If you only update when you really need to, then clearing the whole role cache is a small price to pay. It's simply part of the cost of upgrading Kibana. It's only a problem if you need to do it on every restart.

We certainly can, doing the idempotent PUT on every start-up was supposed to simplify things, but if we're going to have to clear the Role cache when this occurs we can only do this when necessary. 

If we change the logic to only execute the PUT when changes are required, do you still feel that it should be the responsibility of Kibana to explicitly clear the Role cache? Assuming that other applications will begin to use the custom application privileges, I'm unable to think of a scenario where someone would want to update the privileges and not have these changes reflected immediately by the `_has_privileges` API. Perhaps we should always be clearing the Role cache when the privileges are updated in Elasticsearch?

kobelb commented:
So, after thinking about this a bit more, there's one potential issue with checking the privileges and only updating them when they've changed.

Assuming the following pseudo-code:
```
if (!doPrivilegesMatch()) {
  updatePrivileges();
  clearRoleCache();
}
```

If Kibana was to die after the call to `updatePrivileges` and before the call to `clearRoleCache` the next time that Kibana started up, it'd see the privileges as matching and skip clearing the role cache.

Then when we went to check the privileges via the `_has_privilege` API, assuming we're doing the version as a privilege, we'd always get "false".

When the `_has_privilege` check fails with the version action returning false, we could then potentially perform the following pseudo-code:

```
if (!doPrivilegesMatch()) {
  updatePrivileges();
}
 clearRoleCache();
```

where we always clear the role cache.

It does make things more complicated in Kibana, but perhaps it's worth it given the current situation?

tvernum commented:
> We certainly can, doing the idempotent PUT on every start-up was supposed to simplify things, but if we're going to have to clear the Role cache when this occurs we can only do this when necessary.

It's idempotent in the sense that there are no external changes, but it will still issue a write to the security index, and potentially clear caches, etc. We should be hesitant about doing PUTs when we can detect that we don't need to.

> do you still feel that it should be the responsibility of Kibana to explicitly clear the Role cache?

No we can do it in the API.
My concern originally was that if we clear the cache on every API change, that would potentially trigger multiple evictions when a series of PUT/DELETE calls were made. I don't think that's likely to happen, and if we put it into the API the we can apply a finer grained eviction in the future if we find it's a problem.

> If Kibana was to die after the call to updatePrivileges and before the call to clearRoleCache the next time that Kibana started up, it'd see the privileges as matching and skip clearing the role cache.

This is still a problem, even if we do this inside the ES node, but I think we can live with it.
- the cache ought to timeout (but currently doesn't, https://github.com/elastic/elasticsearch/issues/30505)
- the admin can manually clear the cache

> When the _has_privilege check fails with the version action returning false, we could then potentially perform the following pseudo-code ...

The simplest approach I could recommend is to have a single "canary" privilege that is constant between Kibana versions and should be assigned to every user e.g. `action:login`
If a user has that privilege but not the current `version:x.y.z` then something is wrong.
You could then try the privilege PUT again (which will clear the cache) and if that still fails explain that something is wrong.

_BUT_ it means that if you have 2 different (incompatible) Kibana versions running at the same time (for the same application-name / kibana-index) then they'll just round-robin overwriting each others version privilege. It seems to defeat the purpose of using the version privilege as a safefy-gate.

kobelb commented:
> The simplest approach I could recommend is to have a single "canary" privilege that is constant between Kibana versions and should be assigned to every user e.g. action:login
If a user has that privilege but not the current version:x.y.z then something is wrong.
You could then try the privilege PUT again (which will clear the cache) and if that still fails explain that something is wrong.


I think this is a good compromise, and we can only update the privileges when we detect a change is necessary.

> BUT it means that if you have 2 different (incompatible) Kibana versions running at the same time (for the same application-name / kibana-index) then they'll just round-robin overwriting each others version privilege. It seems to defeat the purpose of using the version privilege as a safefy-gate.

I'm afraid this is something that we're going to have to live with, and it should be fine. I'll make sure we're throwing an information error message when this occurs, so it's obvious that there are multiple instances of Kibana running and that's why they aren't able to authorize requests.

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/DeletePrivilegesRequestBuilder.java:
```diff
@@ -0,0 +1,38 @@
+        extends ActionRequestBuilder<DeletePrivilegesRequest, DeletePrivilegesResponse, DeletePrivilegesRequestBuilder>
+        implements WriteRequestBuilder<DeletePrivilegesRequestBuilder> {
+
+    public DeletePrivilegesRequestBuilder(ElasticsearchClient client) {
```
how about we get rid of this constructor and just have the one below and make it public?

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/DeletePrivilegesResponse.java:
```diff
@@ -0,0 +1,67 @@
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        super.readFrom(in);
+        final int size = in.readVInt();
```
Maybe we should just add support for writing a set to StreamOutput/StreamInput?

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/GetPrivilegesRequest.java:
```diff
@@ -0,0 +1,71 @@
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        super.readFrom(in);
+        application = in.readString();
```
this can be null, you need to use `readOptionalString`. Let's add serialization tests for this class too

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/GetPrivilegesRequest.java:
```diff
@@ -0,0 +1,71 @@
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        out.writeString(application);
```
writeOptionalString

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/DeletePrivilegesRequest.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
```
lets add tests around validation

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/DeletePrivilegesRequest.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
```
add tests for serialization

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/DeletePrivilegesResponse.java:
```diff
@@ -0,0 +1,67 @@
+ * Response when deleting application privileges.
+ * Returns a collection of privileges that were successfully found and deleted.
+ */
+public class DeletePrivilegesResponse extends ActionResponse implements ToXContentObject {
```
ideally we should have tests for serialization here too

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/GetPrivilegesAction.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Action for retrieving one or more application privileges from the security index
+ */
+public class GetPrivilegesAction extends Action<GetPrivilegesRequest, GetPrivilegesResponse, GetPrivilegesRequestBuilder> {
```
make it final

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/GetPrivilegesAction.java:
```diff
@@ -0,0 +1,32 @@
+    public static final GetPrivilegesAction INSTANCE = new GetPrivilegesAction();
+    public static final String NAME = "cluster:admin/xpack/security/privilege/get";
+
+    protected GetPrivilegesAction() {
```
private

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/GetPrivilegesRequest.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
```
lets have tests for validation

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/GetPrivilegesRequestBuilder.java:
```diff
@@ -0,0 +1,34 @@
+        this(client, GetPrivilegesAction.INSTANCE);
+    }
+
+    GetPrivilegesRequestBuilder(ElasticsearchClient client, GetPrivilegesAction action) {
```
same thing as the delete action. Lets make this one public and remove the constructor above it. In the client code we can use the action instance

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/GetPrivilegesResponse.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
```
I'd like to have some serialization tests for this

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/PutPrivilegesAction.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Action for putting (adding/updating) one or more application privileges.
+ */
+public class PutPrivilegesAction extends Action<PutPrivilegesRequest, PutPrivilegesResponse, PutPrivilegesRequestBuilder> {
```
make it final

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/PutPrivilegesAction.java:
```diff
@@ -0,0 +1,32 @@
+    public static final PutPrivilegesAction INSTANCE = new PutPrivilegesAction();
+    public static final String NAME = "cluster:admin/xpack/security/privilege/put";
+
+    protected PutPrivilegesAction() {
```
private constructor

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/PutPrivilegesRequest.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
```
lets have tests for serialization

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/PutPrivilegesRequestBuilder.java:
```diff
@@ -0,0 +1,143 @@
+        this(client, PutPrivilegesAction.INSTANCE);
+    }
+
+    PutPrivilegesRequestBuilder(ElasticsearchClient client, PutPrivilegesAction action) {
```
same thing regarding a single constructor

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/PutPrivilegesRequest.java:
```diff
@@ -0,0 +1,101 @@
+/**
+ * Request object to put a one or more application privileges.
+ */
+public class PutPrivilegesRequest extends ActionRequest implements WriteRequest<PutPrivilegesRequest> {
```
make it final

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/GetPrivilegesResponse.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Response containing one or more application privileges retrieved from the security index
+ */
+public class GetPrivilegesResponse extends ActionResponse {
```
make it final

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/GetPrivilegesRequestBuilder.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Builder for {@link GetPrivilegesRequest}
+ */
+public class GetPrivilegesRequestBuilder
```
make it final

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/PutPrivilegesResponse.java:
```diff
@@ -0,0 +1,57 @@
+ * Response when adding one or more application privileges to the security index.
+ * Returns a collection of the privileges that were created (by implication, any other privileges were updated).
+ */
+public class PutPrivilegesResponse extends ActionResponse implements ToXContentObject {
```
make it final

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/PutPrivilegesResponse.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
```
can we add serialization tests?

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/HasPrivilegesRequest.java:
```diff
@@ -85,6 +108,9 @@ public void readFrom(StreamInput in) throws IOException {
         for (int i = 0; i < indexSize; i++) {
             indexPrivileges[i] = RoleDescriptor.IndicesPrivileges.createFrom(in);
         }
+        if (in.getVersion().onOrAfter(Version.V_7_0_0_alpha1)) {
```
can we add serialization tests

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/ApplicationPermission.java:
```diff
@@ -0,0 +1,104 @@
+            this.resources = resources;
+        }
+
+        public boolean grants(ApplicationPrivilege other, Automaton resource) {
```
this can be private since the class is private

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/privilege/TransportPutPrivilegesAction.java:
```diff
@@ -0,0 +1,54 @@
+            listener.onResponse(new PutPrivilegesResponse(Collections.emptyMap()));
+            return;
+        }
+        this.privilegeStore.putPrivileges(request.getPrivileges(), request.getRefreshPolicy(), ActionListener.wrap(
```
just put this in an else and remove the return?

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/user/TransportHasPrivilegesAction.java:
```diff
@@ -66,15 +74,32 @@ protected void doExecute(HasPrivilegesRequest request, ActionListener<HasPrivile
+    private Set<String> getApplicationNames(HasPrivilegesRequest request) {
+        return Arrays.stream(request.applicationPrivileges())
+            .map(RoleDescriptor.ApplicationResourcePrivileges::getApplication)
+            .distinct()
```
distinct and toSet are redundant?

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/user/TransportHasPrivilegesAction.java:
```diff
@@ -66,15 +74,32 @@ protected void doExecute(HasPrivilegesRequest request, ActionListener<HasPrivile
-        logger.debug(() -> new ParameterizedMessage("Check whether role [{}] has privileges cluster=[{}] index=[{}]",
-                Strings.arrayToCommaDelimitedString(userRole.names()), Strings.arrayToCommaDelimitedString(request.clusterPrivileges()),
-                Strings.arrayToCommaDelimitedString(request.indexPrivileges())));
+        logger.debug(() -> new ParameterizedMessage("Check whether role [{}] has privileges cluster=[{}] index=[{}] application=[{}]",
```
this feels like a trace log to me

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -0,0 +1,283 @@
+    }
+
+
+    Collector<Tuple<String, String>, ?, Map<String, List<String>>> tuplesToMap() {
```
make it a private static final instance?

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java:
```diff
@@ -0,0 +1,283 @@
+                                   ActionListener<IndexResponse> listener) {
+        if (privilege.name().size() != 1) {
+            listener.onFailure(new IllegalArgumentException("Cannot store application privileges with multivariate names"));
+            return;
```
use an else instead of return?

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestDeletePrivilegesAction.java:
```diff
@@ -0,0 +1,74 @@
+                });
+    }
+
+    public String[] getPrivileges(RestRequest request) {
```
make it static and private?

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestDeletePrivilegesAction.java:
```diff
@@ -0,0 +1,74 @@
+
+    public String[] getPrivileges(RestRequest request) {
+        final String[] privilegeArray = request.paramAsStringArray("privilege", null);
+        if (privilegeArray == null || (privilegeArray.length == 1 && Strings.isNullOrEmpty(privilegeArray[0]))) {
```
typically we delegate to the action validation

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/AuthorizationServiceTests.java:
```diff
@@ -535,17 +551,17 @@ public void testDenialForAnonymousUser() {
 
         RoleDescriptor role = new RoleDescriptor("a_all", null,
-                new IndicesPrivileges[] { IndicesPrivileges.builder().indices("a").privileges("all").build() }, null);
+            new IndicesPrivileges[]{IndicesPrivileges.builder().indices("a").privileges("all").build()}, null);
```
:(. Was this an autoformatter? I prefer to avoid make these types of changes in an already large PR in addition to my preference for spaces

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java:
```diff
@@ -0,0 +1,275 @@
+    }
+
+    private void validate(boolean validateNames) {
+        if (Regex.isSimpleMatchPattern(application) == false) {
```
We should. It all got a bit ugly somewhere along the way.
 
This exists, purely so `superuser` can have _all_ privileges on _all_ applications.
`SUPERUSER_ROLE_DESCRIPTOR` has `application("*").privileges("*").resources("*")`

Checks in `PutPrivilegesRequest` and `HasPrivilegesRequest` prevent applications wildcards being used in privilege definitions, but they are possible in roles.

The problem really comes from using a single class for defining privileges themselves, and for representing the privileges that a role might grant. You can't define a privilege with a wildcard application, but you can grant one.

I've put in a fix for this check, but the whole thing still feels ugly.

jaymode replied:
> The problem really comes from using a single class for defining privileges themselves, and for representing the privileges that a role might grant. You can't define a privilege with a wildcard application, but you can grant one.

Based on your response I don't feel like you're happy with how this is implemented. How do you feel about us continuing to review this PR and instead of merging to master we start a feature branch where these classes can be separated?

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java:
```diff
@@ -0,0 +1,275 @@
+        return PARSER.parse(parser, allowType);
+    }
+
+    public static ApplicationPrivilege readFrom(StreamInput in) throws IOException {
```
> Can we have a constructor that takes a StreamInput instead of the static method?

Not easily (or at least, not with code that ends up being readable).
The constructors need to call `super` with parameters that are read from the stream. 
There isn't a nice pattern for doing that in a constructor.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/user/TransportHasPrivilegesAction.java:
```diff
@@ -139,4 +196,17 @@ private static boolean testIndex(Automaton checkIndex, Automaton roleIndex) {
+    private boolean testResourceMatch(String application, String checkResource, String checkPrivilegeName, Role userRole,
+                                      Collection<ApplicationPrivilege> privileges) {
+        final Set<String> nameSet = Collections.singleton(checkPrivilegeName);
+        final ApplicationPrivilege checkPrivilege = ApplicationPrivilege.get(application, nameSet, privileges);
```
`ApplicationPrivilege.get` does some *O(n)* on the privileges arg, and `testResourceMatch` is called in a loop.
Do you think you can pull the grouping by privilege name out of the loop?

tvernum replied:
This should be much easier once we have a separate class for privileges stored in the index, so I'm going to hold off on this until then.
There's definitely some major performance improvements we could do, but the API will be much simpler with a better object model (e.g. privileges will always have single name)

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/user/TransportHasPrivilegesAction.java:
```diff
@@ -139,4 +196,17 @@ private static boolean testIndex(Automaton checkIndex, Automaton roleIndex) {
+        assert checkPrivilege.name().equals(nameSet)
+            : "Privilege " + checkPrivilege + " should have name " + nameSet;
+
+        return userRole.application().grants(checkPrivilege, checkResource);
```
Pertaining to the 'Role cache clean saga':
I understand the `checkPrivilege` is freshly baked (not cached), but the `userRole` application permissions are built from cached application privileges. Since we already pull the latest privileges here, do you think it would be possible to refactor `Role` so that it only holds the names of permissions.
I haven't fully grasped all the machinery so this might be nonsense...

tvernum replied:
I think it might be possible.
At the moment we've definitely got the wrong tradeoffs. We're loading privileges from the index every time, and compiling the automata for each of them, but then using the cahced copy on the role.
I'll come back to this after the refactor, because that will change things a bit (the stuff we load from the index won't be an actual privilege object)

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java:
```diff
@@ -0,0 +1,295 @@
+        Set<String> patterns = new HashSet<>();
+        for (String name : names) {
+            name = name.toLowerCase(Locale.ROOT);
+            if (isValidPrivilegeName(name) != false) {
```
nit

albertzaharovits commented:
@tvernum I think this is cleaner
```
POST _xpack/security/privilege
{
  "kibana": {
    "read":{
      "actions":[
        "version:7.0.0-alpha1",
        "action:saved-objects/config/get"
      ]
    }
  }
}
```
than:
```
POST _xpack/security/privilege
{
  "kibana": {
    "read":{
      "application":"kibana",
      "name":"read",
      "actions":[
        "version:7.0.0-alpha1",
        "action:saved-objects/config/get"
      ]
    }
  }
}
```
Curios about the current syntax..

I haven't fully grasped this PR, and I apologies in advance if any of my comments challenge design decisions made long ago. I admit I haven't paid attention from the onset.

tvernum commented:
>  I apologies in advance if any of my comments challenge design decisions made long ago.

No, actually you ask a good question.
The reason is historical - originally the API just took an array as input, but we don't really do array based POSTs in ES (and rest test can't handle it), so I decided it was a better fit to change it to be object based, per the above syntax, but didn't change the rest of the JSON.

I think we can at least make the application and name optional.

tvernum commented:
@jaymode @albertzaharovits 
This is ready for another review.

@albertzaharovits I addressed your comment on the JSON input, but I'm going to hold off on your two  
performance related suggestions until after I split the `ApplicationPrivilege` class.
My plan is to do something like we do for Roles, where the `Role` class represents the runtime behaviour and the `RoleDescriptor` is the stored format.

kobelb commented:
@tvernum I just came across an interesting behavior, when I create a privilege with a capital letter in it, then the `_has_privilege` check always comes back false, but if change it to be all lower case letters then it works correctly.

jaymode commented on server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -969,6 +970,16 @@ public void writeList(List<? extends Writeable> list) throws IOException {
     }
 
+    /**
+     * Writes a list of generic objects via a {@link Writer}
```
s/list/collection

jaymode commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/privilege/GetPrivilegesResponseTests.java:
```diff
@@ -0,0 +1,49 @@
+        assertThat(copy.privileges(), Matchers.equalTo(original.privileges()));
+    }
+
+    private <T> T[] randomArray(int max, IntFunction<T[]> arrayConstructor, Supplier<T> valueConstructor) {
```
lets put this into `ESTestCase` with javadocs and remove the duplication

tvernum commented:
test this please

tvernum commented:
>  when I create a privilege with a capital letter in it, then the _has_privilege check always comes back false
I've tracked down the cause of this. I have a fix for it in a followup PR.
